### PR TITLE
feature: developer service

### PR DESCRIPTION
**Pull request by xaun**
- adds developer provider with the ability to stub widgets templates
- adds specs for the developer provider, and updates to the dashboard svc
- adds specs for widgets service #update method
- adds config for developer provider in workspace/index.js
- adds comments and some restructuring in workspace/index.js 

@cesar-tonnoir please review

### CODE REVIEW

File: DEVELOPER.md [hunk 1]
```diff
@@ -67,36 +67,60 @@ The `workspace/index.js` file is then loaded into `workspace/index.html` via `<s
 
 ##### Process
 
-1. Define the widget's template. The base widgets templates list is retrieved from the MNO API, within each dashboard object. Each widget template will have a structure similar to the following:
+1. **Defining the Widgets Template.**<br>
+  *Widgets templates are currently kept in the maestrano api. They declare defining attributes for each widget. *
+  *It is important to take not of the `path` vs `path` & `metadata.template` attributes. Defining a `metadata.template` enables you to use an existing Impac API engine, and point the front-end to a different template*
```
Comment (cesar-tonnoir) [inline DEVELOPER.md]: typos:
- "not**e**"
- point**s**

Is it mentioned somewhere that by default, "path" will correspond to both the engine path and the template name?

File: src/components/dashboard/dashboard.directive.coffee [hunk 2]
```diff
@@ -10,7 +10,11 @@ module.controller('ImpacDashboardCtrl', ($scope, $http, $q, $filter, $modal, $lo
     # references to services (bound objects shared between all controllers)
     # -------------------------------------
     $scope.currentDhb = ImpacDashboardsSvc.getCurrentDashboard()
-    $scope.widgetsList = ImpacDashboardsSvc.getWidgetsTemplates()
+    # developers can add widgets via the ImpacDeveloper service without modifying the API
+    if ImpacDeveloper.getStatus()
```
Comment (cesar-tonnoir) [inline src/components/dashboard/dashboard.directive.coffee]: So, only the stubbed templates will populate scope.widgetsList?
What happens if the developer loads an existing dashboard with widgets which won't find their templates?

Couldn't we merge the stubbed one into the full list?

File: DEVELOPER.md (hunk 1 above)
Comment (xaun) [inline DEVELOPER.md]: ya widget templates response example comments

File: src/services/dashboards/dashboards.svc.coffee [hunk 3]
```diff
@@ -156,11 +155,18 @@ angular
       )
 
 
+    # Will be filled only once
```
Comment (cesar-tonnoir) [inline src/services/dashboards/dashboards.svc.coffee]: this comment was to explain line `return false if _.isEmpty(templatesArray) || !_.isEmpty(_self.config.widgetsTemplates)`, not the all method

File: src/services/developer/developer.svc.coffee [hunk 4]
```diff
@@ -0,0 +1,45 @@
+#=======================================
```
Comment (cesar-tonnoir) [inline src/services/developer/developer.svc.coffee]: There I see that the templates list is stubbed, but how do you handle the case where a developer tries to create/delete/update "new" widgets and receive an error from Mno API?
Is that stubbed somewhere else?

**Comment by xaun** [comment]
@cesar-tonnoir please review this when you get a chance, cheers

File: src/services/widgets/widgets.svc.coffee [hunk 5]
```diff
@@ -86,12 +92,10 @@ angular
             if needContentReload
               _self.show(updatedSettingsWidget).then(
                 (updatedContentWidget) ->
-                  updatedContentWidget.isLoading = false
                   deferred.resolve(updatedContentWidget)
                 (error) ->
-                  updatedSettingsWidget.isLoading = false
                   deferred.reject(error)
-              )
+              ).finally(-> updatedSettingsWidget.isLoading = false)
```
Comment (xaun) [inline src/services/widgets/widgets.svc.coffee]: @cesar-tonnoir is that ok? it was merged from arnaud's branch

Comment (cesar-tonnoir) [inline src/services/widgets/widgets.svc.coffee]: lgtm :+1:

**Pull request merged**